File: java/src/dsclient/scheduling_policies/MinimalCost.java
package dsclient.scheduling_policies;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import dsclient.JobDispatchPolicy;
import dsclient.JobScheduler;
import dsclient.enums.ServerState;
import dsclient.models.JobSubmission;
import dsclient.models.Resource;
import dsclient.models.ServerType;

public class MinimalCost implements JobDispatchPolicy {
    public void dispatch(JobScheduler scheduler) throws IOException {
        List<ServerType> serverTypes = new ArrayList<>(scheduler.servers);
        Collections.sort(serverTypes, Comparator.comparingDouble(srv -> srv.rate));
        Map<String, Integer> serverOrder = IntStream
                .range(0, serverTypes.size())
                .boxed()
                .collect(Collectors.<Integer, String, Integer>toMap(
                    i -> serverTypes.get(i).type,
                    i -> i
                )
        );

        while (true) {
            String received = scheduler.inquire("REDY");
            if (received.equals("NONE")) {
                break;
            }

            JobSubmission job = JobSubmission.fromReceivedLine(received);

            List<Resource> resources = scheduler.fetchCapableResources(job);

            Resource best = resources.stream().max(
                Comparator.<Resource>comparingInt(
                    res -> -serverOrder.get(res.type)
                ).thenComparingInt(
                    res -> (res.state == ServerState.IDLE) ? 1 : 0
                ).thenComparingInt(
                    res -> (res.state == ServerState.ACTIVE) ? 1 : 0
                )
            ).get();

            scheduler.schedule(job, best);
        }
    }
}

File: python/ds_client/scheduling_policies/minimal_cost.py

from __future__ import annotations
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from ..job_scheduler import JobScheduler

from ..job_dispatch_policy import JobDispatchPolicy
from ..models import JobSubmission
from ..enums import ServerState

class MinimalCost(JobDispatchPolicy):
    def dispatch(self, scheduler: JobScheduler) -> None:
        server_types = sorted(scheduler.servers, key=lambda srv: srv.rate)
        server_order = {srv.type: i for i, srv in enumerate(server_types)}

        while True:
            received = scheduler.inquire('REDY')
            if received == 'NONE':
                break

            job = JobSubmission.from_received_line(received)

            resources = scheduler.fetch_capable_resources(job)

            best = max(
                resources,
                key=lambda res: (
                    -server_order[res.type],
                    res.state == ServerState.IDLE,
                    res.state == ServerState.ACTIVE,
                ),
            )

            scheduler.schedule(job, best)
